compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp

```diff
@@ -49,6 +52,79 @@ struct HostDefaults {
   }
 };
 
+void ResolveCPUAndCPUFeatures(llvm::StringRef inCpu,
+                              llvm::StringRef inCpuFeatures,
+                              const llvm::Triple &triple, std::string &outCpu,
+                              std::string &outCpuFeatures) {
+  // Resolve "host"
+  if (inCpu == "host" || inCpuFeatures == "host") {
+    // If either Cpu or CpuFeatures is "host", the other must be either also
+    // host or the default value.
+    if (!inCpu.empty() && inCpu != "generic" && !inCpuFeatures.empty() &&
+        inCpuFeatures != "host") {
+      llvm::report_fatal_error(
+          "If either Cpu or CpuFeatures is `host`, the other must be either "
+          "also `host` or the default value.",
+          false);
+    }
+    const HostDefaults &hostDefaults = HostDefaults::get();
+    outCpu = hostDefaults.cpu;
+    outCpuFeatures = hostDefaults.cpuFeatures;
+  } else {
+    outCpu = inCpu;
+    outCpuFeatures = inCpuFeatures;
+  }
+
+  // Target-specific CPU feature tweaks that we need unconditionally.
+  if (triple.isAArch64()) {
+    llvm::SubtargetFeatures targetCpuFeatures(outCpuFeatures);
+    // x18 is platform-reserved per the Aarch64 procedure call specification.
+    targetCpuFeatures.AddFeature("reserve-x18", true);
+    outCpuFeatures = targetCpuFeatures.getString();
+  }
+
+  // If CPU is non-host and non-generic then we need to populate the
+  // corresponding features.
+  if (inCpu == "host" || inCpu == "generic" || inCpu.starts_with("generic-")) {
+    return;
+  }
+  if (triple.isX86()) {
+    llvm::SubtargetFeatures targetCpuFeatures(outCpuFeatures);
+    llvm::SmallVector<llvm::StringRef> cpuFeatureList;
+    llvm::X86::getFeaturesForCPU(outCpu, cpuFeatureList);
+    for (auto &feature : cpuFeatureList) {
+      targetCpuFeatures.AddFeature(feature);
+    }
+    outCpuFeatures = targetCpuFeatures.getString();
+    return;
+  }
+  if (triple.isAArch64()) {
+    std::vector<std::string> UpdatedFeaturesVec;
+    std::optional<llvm::AArch64::CpuInfo> cpuInfo =
+        llvm::AArch64::parseCpu(outCpu);
+    if (!cpuInfo) {
+      llvm::report_fatal_error(
+          "Failed to resolve AArch64 CPU features for the specified CPU.",
+          false);
```

Comment:
Let's fix that then. We have held off for too long from providing decent user flag validation and error-handling, resulting in much time wasted for users, such as in this instance, @mariecwhite , but there have been countless other occasions where users have been confused by CPU and CPU feature flags behavior.   So this time around, I will argue more forcefully than I did in earlier instances for prioritizing immediate user-facing validation over internal design considerations.

Reply:
nah, we can't do this - we already today pass in this information via attributes and we can't be crashing python processes hosting the compiler because a user fat-fingered a string.

this should log out errors as normal (llvm::errs) and return a LogicalResult or use InFlightDiagnostics-style result handling to propagate the messages up (ideal, but I haven't done it and usually just errs()).

that we do validation in the LLVMTarget constructor is the issue, and we should change the constructor to just store the inputs and then have a separate `LogicalResult resolve()` or something that does this logic and is able to return failures.

see `LLVMTarget::loadFromConfigAttr` for what is being done today - since that doesn't go through the constructor none of this new logic is running but it should be - loadFromConfigAttr should try the resolve and then return {} as it does for other error cases

Reply:
how about this now?

Reply:
hmm not sure spreading the checks to accesses is great - hiding the LLVMTarget constructor and making sure that they can only be created resolved would make the interface much simpler and prevent the inconsistent/unresolved state

so private ctor and then a `static std::optional<LLVMTarget> makeTarget(...)` (as with what loadFromConfigAttr already does by returning an optional target)

Reply:
ah yes, much better indeed.

Reply:
ok, PTAL @benvanik